overlord/devicestate/devicestate.go

```diff
@@ -333,34 +342,60 @@ func remodelTasks(ctx context.Context, st *state.State, current, new *asserts.Mo
 	}
 	// add new kernel
 	if current.Kernel() != new.Kernel() {
-		// TODO: we need to support corner cases here like:
-		//  0. start with "old-kernel"
-		//  1. remodel to "new-kernel"
-		//  2. remodel back to "old-kernel"
-		// In step (2) we will get a "already-installed" error
-		// here right now (workaround: remove "old-kernel")
-		ts, err := snapstateInstallWithDeviceContext(ctx, st, new.Kernel(), &snapstate.RevisionOptions{Channel: new.KernelTrack()}, userID, snapstate.Flags{}, deviceCtx, fromChange)
+		needsInstall, err := notInstalled(st, new.Kernel())
 		if err != nil {
 			return nil, err
 		}
-		tss = append(tss, ts)
+		if needsInstall {
+			ts, err := snapstateInstallWithDeviceContext(ctx, st, new.Kernel(), &snapstate.RevisionOptions{Channel: new.KernelTrack()}, userID, snapstate.Flags{}, deviceCtx, fromChange)
+			if err != nil {
+				return nil, err
+			}
+			tss = append(tss, ts)
+		} else {
```

Comment:
Nice! This is indeed shorter and easier to read. Updated.

Reply:
Thank you! One note, with this the existing code for gadget below shadows ts with
```
ts, err := snapstateUpdateWithDeviceContext
```

Not strictly an issue right now, but perhaps to avoid nasty suprises it would be better to declare `var ts *state.TaskSet` twice - separately inside `if current.Kernel() != new.Kernel() {...` and in `if current.Base() != new.Base() { ...` blocks.